torch/csrc/api/include/torch/data/ctf/ctf_parser.h

```diff
@@ -0,0 +1,931 @@
+#pragma once
+
+#include <torch/data/ctf/reader.h>
+#include <torch/data/ctf/utils.h>
+
+#include <stdint.h>
+#include <algorithm>
+#include <cassert>
+#include <climits>
+#include <iostream>
+#ifdef CTF_DEBUG
+#include <map>
+#endif
+#include <memory>
+#include <ostream>
+#include <unordered_map>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace ctf {
+
+/*
+ * CTF general format
+ * [Sequence_Id](Sample or Comment)+
+ *   where
+ *          sequence_Id=(empty|[0-9]+)
+ *          Sample=|Input_Name (Value )*
+ *          Comment=|# some content
+ * Example:
+ * 100 |a 1 2 3 |b 100 200
+ * 100 |a 4 5 6 |b 101 201
+ * 100 |b 102983 14532 |a 7 8 9
+ * 100 |a 7 8 9
+ * 200 |b 300 400 |a 10 20 30
+ * 333 |b 500 100
+ * 333 |b 600 -900
+ * 400 |a 1 2 3 |b 100 200
+ * |a 4 5 6 |b 101 201
+ * |a 4 5 6 |b 101 201
+ * 500 |a 1 2 3 |b 100 200
+ */
+
+///
+/// Beginning of type definitions
+///
+
+///
+/// Enumeration type denoting data type of symbolic data entities or actual
+/// data.
+///
+enum class CTFDataType : unsigned int {
+  Unknown = 0,
+  Float = 1,
+  Double = 2,
+  UChar = 3, // So far only used internally in deserializers.
+  Float16 = 4,
+  Int8 = 5,
+  Int16 = 6,
+  Int32 = 7,
+};
+
+///
+/// Enumeration type denoting the format of storage
+///
+enum class CTFDataStorage { Dense, Sparse };
+enum class CTFInputStreamType { Feature, Label };
```

Comment:
I thought it is possible to have more different types other than Feature and Label? If that's the case, does the user need to extend this enum class?

Reply:
One example (CTFSequence) can have more than one input (ctf input stream) that is CTFInputStreamType::Feature and more than one inpu that is CTFInputStreamType::Label

0 |Feature1 1 2 3 |Feature2 3 4 |Label1 10 |Label2 123:456

In the example above, Feature1, Feature2 and Feature3 are input streams that have type=CTFInputStreamType::Feature and Label1 and Label2 have type=CTFInputStreamType::Label

Reply:
Does this mean, we are going to merge feature1 and feature2's data, and return it as a whole as feature data?

Reply:
0 |Feature1 1 2 3 |Feature2 3 4 |Label1 10
0 |Feature1 10 20 30 |Feature2 30 40 |Label1 100

In the case above, Feature1 will be merged (1 2 3 10 2 30), both Feature2 will be merged (3 4 30 40) and Label1 will be merged (10 100)